### PR TITLE
[Dominic Lim] iP

### PR DESCRIPTION
**Pull request by domlimm**
# Week 3's Individual Project Task
> P.S. It's my first PR in CS2103. 😄
<hr>

< **_HELLO THERE_** > if you're reviewing my PR! Hope you're well & thank you for doing this! 🍻 

## Contents of this PR's description
- A numbered list i.e. Features included in PR
- Code block to show a message
- A task list, like Duke!
- Tbh, I'm just filling this list up :')

Aite moving on...

Features available after merge:
1. Greet, Echo, Exit (Bye 👋🏻)
2. Add, List
3. Mark as Done
4. Todo, Event, Deadline
5. Handle Errors
6. ~~Delete~~ 🗑️ 
7. Okay imma stop here...

Now here is a list of things I need to complete in this PR:
- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [ ] a task list
- [x] an emoji
- [x] a blockquote
- [x] a hyper link
- [x] inline code
- [x] some text formatting: bold, italic, stikethrough etc.

Also, just for an `inline code`, don't `git push origin <your_branch_name> --force` please! :')

Now, a Java code block:
```java
while (true) {
    System.out.println("I love CS2103!");
}
```

Alright! Let's wrap up! Thanks again for reviewing my PR!

Feel free to reach out!
[GitHub](https://github.com/domlimm) - [LinkedIn](https://www.linkedin.com/in/dominiclimkj/)

🥂

### CODE REVIEW

File: src/main/java/duke/commands/FindCommand.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                if (i + 1 != length) {
+                    sb.append("\n");
+                }
+            }
+
+            ui.output(sb.toString());
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
```
Comment (bryans17) [inline src/main/java/duke/commands/FindCommand.java]: I like the use of line breaks to separate logical units. Makes the code a lot more readable! Good job!

File: src/main/java/duke/commands/AddCommand.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Adds task to list based on user input.
+ */
+public class AddCommand extends Command {
+    private String fullCommand;
+
+    /**
+     * Creates new AddCommand object that receives user input.
+     * @param fullCommand fullCommand holds contents of the task to be added.
+     */
```
Comment (bryans17) [inline src/main/java/duke/commands/AddCommand.java]: Perhaps insert an empty line between the description and parameter section.

File: src/main/java/duke/commands/Command.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * Requires individual command that extends this class to implement said methods.
+ */
+public abstract class Command {
+    /**
+     * Executes the individual command.
+     * @param taskList taskList holds the list of tasks and its Create, Retrieve, Update, Delete functions.
+     * @param ui ui displays feedback to the user.
+     * @param storage storage handles storage reads and writes.
+     * @throws DukeException Throw when an error occurs during command execution.
+     */
```
Comment (bryans17) [inline src/main/java/duke/commands/Command.java]: Perhaps insert an empty line between the description and parameter section. I noticed the same issue in several other places too.

**Review by bryans17 — COMMENTED** [review]
Overall, a very good job following the Java Coding Standard. I really like the way you use line breaks to separate logical units in your code. Your code is very readable and clean. Only 1 minor issue with JavaDoc comments not having an empty line between the description section and parameters section.

I couldn't find any other Java Coding Standard mistakes. And once again, good job! 👍

**Comment by domlimm** [comment]
Hey there @bryans17 ! Thank you so much for reviewing my PR and pointing out my mistake! Overlooked on that part.

I have updated all javadocs as per your suggestion and more importantly, the code standard.

Thanks so much! Happy CNY! 🍻

File: src/main/java/duke/commands/Command.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package duke.commands;
```
Comment (tiewweijian) [inline src/main/java/duke/commands/Command.java:1]: Great way of packaging your classes together. Very neat and readable.

File: src/main/java/duke/commands/ToggleCommand.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    ? "Nice! I've marked this task as done:\n"
+                    : "OK, I've marked this task as not done yet:\n";
+
+            ui.output(output + taskString);
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException(Constants.INVALID_INDEX_MSG);
+        } catch (IOException e) {
+            throw new DukeException(Constants.STORAGE_UPDATE_MSG);
+        } catch (IllegalArgumentException e) {
+            throw new DukeException(Constants.INVALID_MARK_MSG);
```
Comment (tiewweijian) [inline src/main/java/duke/commands/ToggleCommand.java:101]: Great naming conventions to your constants.

File: src/main/java/duke/storage/Storage.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import java.nio.file.Paths;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import duke.constants.Constants;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
```
Comment (tiewweijian) [inline src/main/java/duke/storage/Storage.java:18]: Great way to separate your imports.

File: src/main/java/duke/task/Event.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * Gets the date of event of this task.
+     * @return Returns the date of event of this task.
+     */
+    public LocalDate getDate() {
+        return this.date;
+    }
+
+    /**
+     * Returns a formatted string of an Event task.
```
Comment (tiewweijian) [inline src/main/java/duke/task/Event.java:38]: You might have incorrectly formatted the javadocs for this method. 
Perhaps it should look something like this
```
/**
 * @returns a formatted string of an Event task. 
 */
```

**Review by tiewweijian — COMMENTED** [review]
I like how you format your codes and how you organise them into packages. Very intuitive and easy to read. 👍🏻  Keep up the good work! 🥂

**Comment by domlimm** [comment]
Hello @tiewweijian ! I appreciate the inputs! Thank you so much for reviewing my PR!

File: src/main/java/duke/task/Event.java (hunk 7 above)
Comment (domlimm) [reply src/main/java/duke/task/Event.java:38]: I apologise, this is slightly misleading! I should have documented both the description and the return value. Thanks for catching it!

File: src/main/java/duke/commands/AddCommand.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            type = taskArr[0];
+
+            if (type.equalsIgnoreCase("deadline")) {
+                String[] taskData = taskArr[1].split(" /by ");
+                LocalDate by = LocalDate.parse(taskData[1]);
+
+                task = new Deadline(taskData[0], by);
+                textToAdd = String.format(
+                        "D | %s | %s | %s",
+                        task.getCompleted() ? "1" : "0",
```
Comment (brandonrhan) [inline src/main/java/duke/commands/AddCommand.java]: maybe you could rename this task.getCompleted() function

File: src/main/java/duke/commands/ListCommand.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Executes the list command.
+     */
+    @Override
+    public String execute(TaskList taskList, Storage storage) {
+        String response = "";
+        int length = taskList.size();
+        StringBuilder sb = new StringBuilder();
+
+        if (length == 0) {
```
Comment (brandonrhan) [inline src/main/java/duke/commands/ListCommand.java:25]: this part is repeated for all subclasses of Command, maybe you could move it to Command class?

File: src/main/java/duke/commands/ToggleCommand.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+            }
+
+            if (!Utils.isNumeric(this.args)) {
+                throw new NumberFormatException();
+            }
+
+            int index = Integer.parseInt(this.args);
+            Task updatedTask = taskList.toggleCompleted(this.isMark, --index);
+            String taskString = updatedTask.toString();
+            String updateString = "";
```
Comment (brandonrhan) [inline src/main/java/duke/commands/ToggleCommand.java]: should it be updatedString wich is similar to updatedTask

File: src/main/java/duke/constants/Constants.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
+package duke.constants;
+
+/**
+ * Holds all the immutable strings used for displaying to the user.
+ */
+public final class Constants {
```
Comment (brandonrhan) [inline src/main/java/duke/constants/Constants.java:6]: I really like this Constants class, it is so helpful!

File: src/main/java/duke/parser/Parser.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        } else if (isAdd(command)) {
+            return new AddCommand(fullCommand);
+        } else if (command.equalsIgnoreCase("find")) {
+            return new FindCommand(args);
+        } else {
+            throw new DukeException(Constants.UNKNOWN_MSG);
+        }
+    }
+
+    private boolean isAdd(String command) {
```
Comment (brandonrhan) [inline src/main/java/duke/parser/Parser.java:64]: maybe should rename this function

File: src/main/java/duke/storage/Storage.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        if (!file.exists()) {
+            file.createNewFile();
+        }
+
+        sc = new Scanner(file);
+
+        while (sc.hasNext()) {
+            String task = sc.nextLine();
+            String[] taskData = task.split("\\s\\|\\s");
+            Boolean completed = Integer.parseInt(taskData[1]) == 1;
```
Comment (brandonrhan) [inline src/main/java/duke/storage/Storage.java]: Maybe could rename this boolean as isCompleted

**Review by brandonrhan — COMMENTED** [review]
I really like your code! The logic is very clear and easy to read. Just a suggestion maybe could us enums for different kinds of command. I have learned a lot from this!

**Comment by domlimm** [comment]
Hey @brandonrhan ! I appreciate all the inputs. Will make the necessary changes from your suggestions! Appreciate the time you've taken to review my PR, thanks a bunch! 😃

**Pull request closed without merging**